draft-ietf-protected-audience-key-value-services.md

```diff
@@ -0,0 +1,520 @@
+~~~~~ cbor-diag
+[
+  {
+    "id": 0,
+    "keyGroupOutputs": [
+      {
+        "tags": [
+          "interestGroupNames"
+        ],
+        "keyValues": {
+          "InterestGroup1": {
+            "value": "{\"priorityVector\":{\"signal1\":1}}"
+          }
+        }
+      },
+      {
+        "tags": [
+          "keys"
+        ],
+        "keyValues": {
+          "keyAfromInterestGroup1": {
+            "value": "valueForA"
+          },
+          "keyBfromInterestGroup1": {
+            "value":"[\"value1ForB\",\"value2ForB\"]"
+          }
+        }
+      }
+    ]
+  }
+]
+~~~~~
+
+### Structured keys response specification
+
+Structured keys are keys that the client is aware of and the client
+can use the response to do
+additional processing. The value of these keys must abide by the
+following schema for the client to
+successfully parse them.
+
+Note that they must be serialized to string when stored as the value.
+
+#### InterestGroupResponse
+
+The schema below is defined following the spec by
+https://json-schema.org. For values for keys
+from the `interestGroupNames` namespace, they must conform to the
+following schema, prior to being
+serialized to string.
+
+~~~~~ json
+{
+    "title": "tkv.response.v2.InterestGroupResponse",
+    "description": "Format for value of keys in groups tagged 'interestGroupNames'",
+    "type": "object",
+    "additionalProperties": false,
+    "properties": {
+        "priorityVector": {
+            "type": "object",
+            "patternProperties": {
+                ".*": {
+                    "description": "signals",
+                    "type": "number"
+                }
+            }
+        },
+        "updateIfOlderThanMs": {
+            "description": "This optional field specifies that the interest group should be updated if the interest group hasn't been joined or updated in a duration of time exceeding `updateIfOlderThanMs` milliseconds. Updates that ended in failure, either parse or network failure, are not considered to increment the last update or join time. An `updateIfOlderThanMs` that's less than 10 minutes will be clamped to 10 minutes.",
+            "type": "unsigned integer"
+        }
+    }
+}
+~~~~~
+
+Example:
+
+~~~~~ json
+{
+    "priorityVector": {
+        "signal1": 1,
+        "signal2": 2
+    },
+    "updateIfOlderThanMs": 10000
+}
+~~~~~
+
+### Generating a Response
+
+This algorithm describes how the Key Value Service MAY generate a response to a request.
+
+The input is a `payload` corresponding to {{response-schema}} and the HPKE receiver context saved in
+{{request-parsing}}, `rctxt`.
+
+The output is a `response` to be sent to a Client.
+
+1. Let `cbor payload` equal the
+   [deterministically encoded CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c)
+   `payload`. Return an empty `response` on CBOR encoding failure.
+1. Let `compressed payload` equal the [GZIP] compressed `cbor payload`, returning an empty
```

Comment:
This is not correct. We can't compress the top-level response. That leaves us open to compression ratio side-channel attacks (something to also mention in security considerations).

Reply:
Thanks for catching that.